File: app/src/main/java/tech/michaeloverman/mscount/database/WorksListAdapter.java
/* Copyright (C) 2017 Michael Overman - All Rights Reserved */
package tech.michaeloverman.mscount.database;

import android.content.Context;
import android.database.Cursor;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import tech.michaeloverman.mscount.R;
import tech.michaeloverman.mscount.pojos.TitleKeyObject;
import tech.michaeloverman.mscount.utils.AlphanumComparator;
import timber.log.Timber;

/**
 * Adapter for supplying recycler view with proper data.
 * Created by Michael on 2/25/2017.
 */

public class WorksListAdapter extends RecyclerView.Adapter<WorksListAdapter.WorksViewHolder> {

    private final Context mContext;
    private List<TitleKeyObject> mTitles;
    private final WorksListAdapterOnClickHandler mClickHandler;

    public WorksListAdapter(@NonNull Context context, WorksListAdapterOnClickHandler handler) {
        mContext = context;
//        mTitles = titles;
        mClickHandler = handler;
    }

    interface WorksListAdapterOnClickHandler {
        void onClick(String key, String title);
    }

    @NotNull
    @Override
    public WorksViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
//        Timber.d("onCreateViewHolder()");
        View item = LayoutInflater.from(mContext).inflate(R.layout.list_item_work, parent, false);
        return new WorksViewHolder(item);
    }

    @Override
    public void onBindViewHolder(WorksViewHolder holder, int position) {
//        Timber.d("onBindViewHolder()");
        holder.title.setText(mTitles.get(position).getTitle());
        holder.title.setHorizontallyScrolling(true);
        holder.title.setEllipsize(TextUtils.TruncateAt.MIDDLE);
//        ViewCompat.setTransitionName(holder.title, "titleViewTrans" + position);
    }

    @Override
    public int getItemCount() {
        return mTitles == null ? 0 : mTitles.size();
    }

    public void setTitles(List<TitleKeyObject> titles) {
        mTitles = titles;
        //noinspection unchecked
        Collections.sort(mTitles, new AlphanumComparator());
        Timber.d("setTitles() - %s titles...", mTitles.size());
        notifyDataSetChanged();
    }

    public void newCursor(Cursor data) {
        List<TitleKeyObject> titles = new ArrayList<>();
        if(data == null) {
            Timber.d("Null cursor...");
            mTitles = titles;
            return;
        }

        Timber.d("newCursor() data.getCount() == %s", data.getCount());

        try {
            data.moveToFirst();
            while (!data.isAfterLast()) {
                titles.add(new TitleKeyObject(
                        data.getString(ProgramDatabaseSchema.MetProgram.POSITION_TITLE),
                        data.getString(ProgramDatabaseSchema.MetProgram.POSITION_ID)));
                data.moveToNext();
            }
        } catch (Exception exception) {
            Timber.d("Problem here: check this out...");
        }
        setTitles(titles);
    }

    class WorksViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.work_title)
        TextView title;

        WorksViewHolder(View itemView) {
            super(itemView);
//            Timber.d("WorksViewHolder constructor()");
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            Timber.d("WorksViewHolder onClick()");
            int position = getAdapterPosition();
            String key = mTitles.get(position).getKey();
            String title = mTitles.get(position).getTitle();
            mClickHandler.onClick(key, title);
        }
    }
}

File: app/src/main/java/tech/michaeloverman/mscount/database/WorksListAdapter.kt
/* Copyright (C) 2017 Michael Overman - All Rights Reserved */
package tech.michaeloverman.mscount.database

import android.content.Context
import android.database.Cursor
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.list_item_work.view.*
import tech.michaeloverman.mscount.R
import tech.michaeloverman.mscount.database.WorksListAdapter.WorksViewHolder
import tech.michaeloverman.mscount.pojos.TitleKeyObject
import tech.michaeloverman.mscount.utils.AlphanumComparator
import timber.log.Timber
import java.util.*

/**
 * Adapter for supplying recycler view with proper data.
 * Created by Michael on 2/25/2017.
 */
class WorksListAdapter     //        mTitles = titles;
(private val mContext: Context, private val mClickHandler: WorksListAdapterOnClickHandler) : RecyclerView.Adapter<WorksViewHolder>() {
    private var mTitles: List<TitleKeyObject> = emptyList()

    interface WorksListAdapterOnClickHandler {
        fun onClick(key: String?, title: String?)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WorksViewHolder {
//        Timber.d("onCreateViewHolder()");
        val item = LayoutInflater.from(mContext).inflate(R.layout.list_item_work, parent, false)
        return WorksViewHolder(item)
    }

    override fun onBindViewHolder(holder: WorksViewHolder, position: Int) {
//        Timber.d("onBindViewHolder()");
        holder.title.text = mTitles[position].title
        holder.title.setHorizontallyScrolling(true)
        holder.title.ellipsize = TextUtils.TruncateAt.MIDDLE
        //        ViewCompat.setTransitionName(holder.title, "titleViewTrans" + position);
    }

    override fun getItemCount(): Int {
        return mTitles.size
    }

    fun setTitles(titles: List<TitleKeyObject>) {
        mTitles = titles
        Collections.sort(mTitles, AlphanumComparator())
        Timber.d("setTitles() - %s titles...", mTitles.size)
        notifyDataSetChanged()
    }

    fun newCursor(data: Cursor?) {
        val titles: MutableList<TitleKeyObject> = ArrayList()
        if (data == null) {
            Timber.d("Null cursor...")
            mTitles = titles
            return
        }
        Timber.d("newCursor() data.getCount() == %s", data.count)
        try {
            data.moveToFirst()
            while (!data.isAfterLast) {
                titles.add(TitleKeyObject(
                        data.getString(ProgramDatabaseSchema.MetProgram.POSITION_TITLE),
                        data.getString(ProgramDatabaseSchema.MetProgram.POSITION_ID)))
                data.moveToNext()
            }
        } catch (exception: Exception) {
            Timber.d("Problem here: check this out...")
        }
        setTitles(titles)
    }

    inner class WorksViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var title: TextView = itemView.work_title
        override fun onClick(v: View) {
            Timber.d("WorksViewHolder onClick()")
            val position = adapterPosition
            val key = mTitles[position].key
            val title = mTitles[position].title
            mClickHandler.onClick(key, title)
        }

        init {
            //            Timber.d("WorksViewHolder constructor()");
            itemView.setOnClickListener(this)
        }
    }
}